### ISSUE TITLE
Builds are not (fully) reproducible due to file permissions stored in .whl

### ISSUE THREAD

**Issue by miccoli** [pypa/wheel#362]
In order to obtain a fully reproducible build one has to build the wheel with the same `umask`.

Here how to reproduce the issue; tested with wheel 0.34.2, using pypa/sampleproject as an example.

```
$ export SOURCE_DATE_EPOCH=$(git log -n 1 --pretty=%ct)
$ echo $SOURCE_DATE_EPOCH 
1593523015
```

```
$ umask 022
$ python setup.py --quiet bdist_wheel
$ sha3-512sum -N 32 dist/sampleproject-2.0.0-py3-none-any.whl 
A64A8921  dist/sampleproject-2.0.0-py3-none-any.whl
```
but changing the `umask` I have
```
$ umask 000
$ python setup.py --quiet bdist_wheel
$ sha3-512sum -N 32 dist/sampleproject-2.0.0-py3-none-any.whl 
FED67824  dist/sampleproject-2.0.0-py3-none-any.whl
```

This is due to the fact that file permissions, stored in the `.whl` file, are affected by the umask at build time.
```
$ zipinfo dist/sampleproject-2.0.0-py3-none-any.whl
Archive:  dist/sampleproject-2.0.0-py3-none-any.whl
Zip file size: 4208 bytes, number of entries: 10
-rw-rw-rw-  2.0 unx      111 b- defN 20-Jun-30 13:16 sample/__init__.py
-rw-r--r--  2.0 unx        9 b- defN 20-Jun-30 13:16 sample/package_data.dat
-rw-rw-rw-  2.0 unx       43 b- defN 20-Jun-30 13:16 sample/simple.py
-rw-r--r--  2.0 unx        9 b- defN 20-Jun-30 13:16 sampleproject-2.0.0.data/data/my_data/data_file
-rw-r--r--  2.0 unx     1081 b- defN 20-Jun-30 13:16 sampleproject-2.0.0.dist-info/LICENSE.txt
-rw-rw-rw-  2.0 unx     3043 b- defN 20-Jun-30 13:16 sampleproject-2.0.0.dist-info/METADATA
-rw-rw-rw-  2.0 unx       92 b- defN 20-Jun-30 13:16 sampleproject-2.0.0.dist-info/WHEEL
-rw-rw-rw-  2.0 unx       40 b- defN 20-Jun-30 13:16 sampleproject-2.0.0.dist-info/entry_points.txt
-rw-rw-rw-  2.0 unx        7 b- defN 20-Jun-30 13:16 sampleproject-2.0.0.dist-info/top_level.txt
?rw-rw-r--  2.0 unx      843 b- defN 20-Jun-30 13:16 sampleproject-2.0.0.dist-info/RECORD
10 files, 5278 bytes uncompressed, 2740 bytes compressed:  48.1%
```

I think that file permissions should be normalised in the `.whl` file, and not dependant on the build environment. This same approach is implemented in flit, since [v. 0.12](https://flit.readthedocs.io/en/latest/history.html#version-0-12)

**Comment by agronholm** [comment]
> I think that file permissions should be normalised in the .whl file, and not dependant on the build environment

I fully agree, and I've raised this point in discussions with other PyPA people. So far there have been no objections.

**Comment by nanonyme** [comment]
Isn't it a reasonable assumption that if you want reproducible build system, you set umask in it?

**Comment by miccoli** [comment]
> Isn't it a reasonable assumption that if you want reproducible build system, you set umask in it?

May be, but `umask` settings is a little bit of extra information that you have to explicitly set in order to reproduce a build. 
Other build systems (flit, hatch) *do* normalize permission in the wheel file, and I do not see any reason against permission normalization.

May be this can be implemented in a new major release, if there are concerns about back compatibility.

For sure this is not a critical topic, it is just about which is the best practice when constructing `.whl` archives.

**Comment by nanonyme** [comment]
What about executable permission? It is apparently expected wheel will pick up that from workspace. Normalisation can't simply ignore file permissions if done.

**Comment by miccoli** [comment]
> What about executable permission?

I did not define what exactly "normalizing file permissions" means, but I'm not advocating just nuking all permissions. Maybe a sensible approach would be to mimic git and use
- octal 755 for all executable files
- octal 644 otherwise
...

**Comment by agronholm** [comment]
That would not be very reproduceable on Windows.

**Comment by Olindholm** [comment]
For a long time I've been trying to create reproducible builds (wheels). I can't say why I didn't find any information about it earlier. But today I did find out about the `SOURCE_DATE_EPOCH` environment variable. Amazing I thought! Finally building the same thing twice resulted in the same hash.

However, I ran into this problem not soon after when I started trying to build inside docker images. I was fumbling around quite some time before I figured out that it was a permissions discrepancy inside zip files (wheels) between what I build on my host, and what I build inside docker images. A real pain to be honest.

But anyway, seems i got it working really nice now. Inside my `setup.py` I have the following:
```
# Set build date and umask (makes wheels reprocudible)
os.environ["SOURCE_DATE_EPOCH"] = "315532800"  # 1980-01-01 00:00 UTC
os.umask(0o022)
```

**Comment by jayaddison** [comment]
To add some supporting data for this: a recent-ish [paper](https://www.cs.cmu.edu/~ckaestne/pdf/icse25_rb.pdf) presented at [ICSE2025](https://conf.researchr.org/track/icse-2025/icse-2025-research-track) reports that reproducibility of ~~398~~ 4,000 PyPi packages improved from ~12% to ~98% as a result of standardising `umask` and timestamps.  It does note that packages that use `flit` are not affected by the `umask` issue (in fact, the ~12% of reproducible packages seem to be packages that use `flit` or `hatch` as package build mechanism).

Edit: fixup: 4,000 packages, not 398

**Comment by jayaddison** [comment]
(my mistake; the sample size was ~10x that - 4,000 PyPi packages, not 398.  I've edited my comment)

**Comment by jayaddison** [comment]
@agronholm 
> That would not be very reproduceable on Windows.

Is there a way to determine whether a file path destined for a wheel file is a script or not?  If so, I figure that that could be a reasonable way to assign constant file-permission bits (say, `0o664` for most files and `0o775` for script files).

(I guess it is probably not this simple, but I'm going to ask anyway 😄)

**Comment by nanonyme** [comment]
Technically you're not supposed to be shipping scripts since long ago but write entrypoint metadata and have installer generate scripts. I think most do. So while my earlier point is valid it might not be worth it to die on that hill.

**Comment by henryiii** [comment]
This is a build backend issue, and needs to be solved on each backend, not here. `wheel` is not a build backend, it's a CLI. Maybe having some tool here to standardize wheel permissions on an existing wheel would be in scope, though?

**Comment by nanonyme** [comment]
I was just angling that it might be justified to do best effort normalisation if done at all.

**Comment by miccoli** [comment]
When I opened this issue this repo was the canonical location of `bdist_wheel`, so this was the correct place to implement permission normalisation (I think). But things have changed:

https://github.com/pypa/wheel/blob/197012dcb8a9da10570d6486bc1a70305861e7f2/src/wheel/bdist_wheel.py#L4-L10

I just checked with setuptools 82.0.1, and the problem still persists.

I propose to migrate this issue to pypa/setuptools.

**Comment by jayaddison** [comment]
@miccoli I had begun drafting some local changes to `wheel.git` locally to explore an implementation, and did find relevant code and test cases here; for example the [`WheelFile` class definition](https://github.com/pypa/wheel/blob/197012dcb8a9da10570d6486bc1a70305861e7f2/src/wheel/wheelfile.py#L58), and some of the packing/unpacking tests.

The tests include assertions on Zip file permissions, and in particular I was looking at these lines of code that I believe would implicitly inherit permissions onto/into the written file based on the user `umask` at runtime:

https://github.com/pypa/wheel/blob/197012dcb8a9da10570d6486bc1a70305861e7f2/src/wheel/wheelfile.py#L196-L203

Before moving the issue: are we sure the code here is no longer relevant?
